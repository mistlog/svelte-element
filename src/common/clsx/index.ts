/**
 * from https://github.com/lukeed/clsx
 */
function toVal(mix) {
	var k, y, str = '';

	if (typeof mix === 'string' || typeof mix === 'number') {
		str += mix;
	} else if (typeof mix === 'object') {
		if (Array.isArray(mix)) {
			for (k = 0; k < mix.length; k++) {
				if (mix[k]) {
					if (y = toVal(mix[k])) {
						str && (str += ' ');
						str += y;
					}
				}
			}
		} else {
			for (k in mix) {
				if (mix[k]) {
					str && (str += ' ');
					str += k;
				}
			}
		}
	}

	return str;
}

export type ClassValue = ClassArray | ClassDictionary | string | number | null | boolean | undefined;

export interface ClassDictionary {
	[id: string]: any;
}

export interface ClassArray extends Array<ClassValue> { }

export function Î£(...args: ClassArray) {
	var i = 0, tmp, x, str = '';
	while (i < arguments.length) {
		if (tmp = arguments[i++]) {
			if (x = toVal(tmp)) {
				str && (str += ' ');
				str += x
			}
		}
	}
	return str;
}
