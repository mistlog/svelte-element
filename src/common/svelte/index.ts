import { setContext as rawSetContext, getContext as rawGetContext } from "svelte";

export function setContext<T>(key: keyof T, value: T[keyof T]) {
    return rawSetContext(key, value);
}

export function getContext<T>(key: keyof T): T[keyof T] {
    return rawGetContext(key);
}

export type SvelteInternalProps = { $$slots?: any };
