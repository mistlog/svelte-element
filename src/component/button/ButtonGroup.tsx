function UI(style: string) {
    <div class={style}>
        <slot />
    </div>
}

export default function _() {
    //@ts-ignore
    <Style />;

    //@ts-ignore
    <UI />;
}

function Style() {
    const style = "el-button-group";
}