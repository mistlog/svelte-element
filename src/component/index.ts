export { default as Button } from "./button/Button";
export * from "./button/Button";

export { default as ButtonGroup } from "./button/ButtonGroup";

export { default as Row } from "./layout/Row";
export { default as Col } from "./layout/Col";

export { default as Radio } from "./radio/Radio";
export * from "./radio/Radio";
export { default as RadioGroup } from "./radio/RadioGroup";
export * from "./radio/RadioGroup";

export { default as Checkbox } from "./checkbox/Checkbox";
export * from "./checkbox/Checkbox";
export { default as CheckboxGroup } from "./checkbox/CheckboxGroup";
export * from "./checkbox/CheckboxGroup";

export { default as Menu } from "./menu/Menu";
export { default as MenuItem } from "./menu/MenuItem";
export { default as SubMenu } from "./menu/SubMenu";
