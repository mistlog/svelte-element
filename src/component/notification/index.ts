import Notification, { INotificationWillUnmountEvent, NotificationType } from "./Notification";

export interface INotificationConfig {
    title: string;
    message: string;
    duration?: number;
    type?: NotificationType;
}

export function notification(config: INotificationConfig) {
    const { title = "", message = "", duration = 3000, type } = config || {};

    //
    const div = document.createElement("div");
    document.body.appendChild(div);

    const className = ".el-notification";

    // credit: element react
    const instances = document.querySelectorAll(className);
    const lastInstance = instances[instances.length - 1] as HTMLElement;
    const top = lastInstance ? lastInstance.getBoundingClientRect().bottom + 16 : 16;

    let component = Reflect.construct(Notification, [
        {
            target: div,
            props: { top, title, message, type },
            intro: true
        }
    ]);

    //
    const removeNotification = () => {
        if (component) {
            component.$destroy();
            component = null;
        }
    };

    component.$on("close", () => {
        removeNotification();
    });

    component.$on("willUnmount", (e: INotificationWillUnmountEvent) => {
        // credit: element react
        const { height, top } = e.detail;
        const instances = document.querySelectorAll(className);
        for (let i = 0; i < instances.length; ++i) {
            const element = instances[i] as HTMLElement;
            const rect = element.getBoundingClientRect();
            if (rect.top > top) {
                element.style.setProperty("top", (rect.top - height - 16).toString() + "px");
            }
        }
        document.body.removeChild(div);
    });

    setTimeout(() => {
        removeNotification();
    }, duration);
}
