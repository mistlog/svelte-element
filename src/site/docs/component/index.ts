import h1 from "./h1";
import h2 from "./h2";
import h3 from "./h3";

export { h1, h2, h3 };