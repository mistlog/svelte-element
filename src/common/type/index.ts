import { Readable, Writable } from "svelte/store";

export type InputOnChangeHandler = (event: Svelte.ChangeEvent<HTMLInputElement>) => void;
export type ButtonOnClickHandler = (event: Svelte.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

type TypeWithGeneric<T> = Readable<T> | Writable<T>;
export type AutoSubscribe<Type> = Type extends TypeWithGeneric<infer X> ? X : never;

export type IconName = "edit" | "share" | "delete" | "search" | "arrow-left";