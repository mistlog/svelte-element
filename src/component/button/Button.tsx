function UI(state: IButtonState, { loading, icon }: IButtonProps) {
    <button class={state.style} onClick>
        <if condition={loading}>
            <i class={state.loadingStyle} />
        </if>
        <if condition={Boolean(icon)}>
            <i class={state.iconStyle} />
        </if>
        <if condition={state.slotExists}>
            <span>
                <slot />
            </span>
        </if>
    </button>;
}

export interface IButtonProps {
    /**
     * button type
     * @default "default"
     */
    type?: ButtonType;
    /**
     * button size
     * @default ""
     */
    size?: ButtonSize;
    /**
     * disable the button
     * @default false
     */
    disabled?: boolean;
    /**
     * determine whether it's loading
     * @default false
     */
    loading?: boolean;
    /**
     * determine whether it's a plain button,
     * @default false
     */
    plain?: boolean;
    /**
     * @default ""
     */
    icon?: IconName;
    onClick?: ButtonOnClickHandler;
}

export type ButtonType = "primary" | "text" | "default" | "success" | "warning" | "danger" | "info";
export type ButtonSize = "small" | "mini";

/**
 * by convention, the default export would be the component
 * thus the name of it doesn't matter here
 */
export default function _(props: IButtonProps) {
    const {
        type = "default",
        disabled = false,
        plain = false,
        icon = "",
        loading = false,
        size = ""
    } = props;

    //@ts-ignore
    <State />;

    //@ts-ignore
    <UI />;
}

interface IButtonState {
    slotExists: boolean;
    iconStyle: string;
    loadingStyle: string;
    style: string;
}

function State(
    $$props: SvelteInternalProps,
    { icon, type, disabled, plain, loading, size }: IButtonProps
) {
    const state = {} as IButtonState;

    state.slotExists = Boolean($$props.$$slots);
    state.loadingStyle = `el-icon-loading`;

    {
        ("use watch");
        state.iconStyle = `el-icon-${icon}`;
        state.style = Σ("el-button", `el-button--${type}`, {
            "is-disabled": disabled,
            "is-plain": plain,
            "is-loading": loading,
            [`el-button--${size}`]: Boolean(size)
        });
    }
}

import { Σ } from "../../common/clsx/index";
import { IconName, ButtonOnClickHandler } from "../../common/type";
import { SvelteInternalProps } from "../../common/svelte";
