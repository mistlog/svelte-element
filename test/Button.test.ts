import { render, fireEvent } from "@testing-library/svelte";
import { normalize } from "./common/utility";

import { IButtonProps } from "../src/component/index";
import ButtonGroup from "../src/component/button/ButtonGroup.svelte";
import Button from "../src/component/button/Button.svelte";

import DefaultButton from "./assets/DefaultButton.svelte";
import ClickButton from "./assets/ClickButton.svelte";
import TwoButtons from "./assets/TwoButtons.svelte";

describe("Button", () => {
    test("text and default color", () => {
        const { container } = render(DefaultButton);
        const button = container.querySelector("button");
        expect(normalize(button.outerHTML)).toMatchSnapshot();
    });

    test("button type", () => {
        const { container } = render(Button, { type: "primary" } as IButtonProps);
        const button = container.querySelector("button");
        expect(normalize(button.outerHTML)).toMatchSnapshot();
    });

    test("loading", () => {
        const { container } = render(Button, { loading: true } as IButtonProps);
        const button = container.querySelector("button");
        expect(normalize(button.outerHTML)).toMatchSnapshot();
    });

    test("icon", () => {
        const { container } = render(Button, { icon: "edit" } as IButtonProps);
        const button = container.querySelector("button");
        expect(normalize(button.outerHTML)).toMatchSnapshot();
    });

    test("empty button", () => {
        const { container } = render(Button);
        const button = container.querySelector("button");
        expect(normalize(button.outerHTML)).toMatchSnapshot();
    });

    test("click event", async () => {
        const { container } = render(ClickButton);
        const button = container.querySelector("button");
        expect(normalize(button.outerHTML)).toMatchSnapshot();

        await fireEvent(button, new MouseEvent("click", {}));
        expect(normalize(button.outerHTML)).toMatchSnapshot();
    });
});

describe("ButtonGroup", () => {
    test("empty button group", () => {
        const { container } = render(ButtonGroup);
        const buttonGroup = container.querySelector(".el-button-group");
        expect(normalize(buttonGroup.outerHTML)).toMatchSnapshot();
    });

    test("button group with buttons", () => {
        const { container } = render(TwoButtons);
        const buttonGroup = container.querySelector(".el-button-group");
        expect(normalize(buttonGroup.outerHTML)).toMatchSnapshot();
    });
});
