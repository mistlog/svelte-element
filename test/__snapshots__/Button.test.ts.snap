// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button button type 1`] = `
"<button class=\\"el-button el-button--primary\\"></button>
"
`;

exports[`Button click event 1`] = `
"<button class=\\"el-button el-button--primary\\"><span>Load</span></button>
"
`;

exports[`Button click event 2`] = `
"<button class=\\"el-button el-button--default is-loading\\">
  <i class=\\"el-icon-loading\\"></i>
  <span>Loading...</span>
</button>
"
`;

exports[`Button empty button 1`] = `
"<button class=\\"el-button el-button--default\\"></button>
"
`;

exports[`Button icon 1`] = `
"<button class=\\"el-button el-button--default\\">
  <i class=\\"el-icon-edit\\"></i>
</button>
"
`;

exports[`Button loading 1`] = `
"<button class=\\"el-button el-button--default is-loading\\">
  <i class=\\"el-icon-loading\\"></i>
</button>
"
`;

exports[`Button text and default color 1`] = `
"<button class=\\"el-button el-button--default\\">
  <span>Default Button</span>
</button>
"
`;

exports[`ButtonGroup button group with buttons 1`] = `
"<div class=\\"el-button-group\\">
  <button class=\\"el-button el-button--primary\\">
    <i class=\\"el-icon-edit\\"></i>
  </button>
  <button class=\\"el-button el-button--primary\\">
    <i class=\\"el-icon-share\\"></i>
  </button>
</div>
"
`;

exports[`ButtonGroup empty button group 1`] = `
"<div class=\\"el-button-group\\"></div>
"
`;
